AX: WebKit sometimes exposes duplicate text for unrendered text nodes (e.g. labels inside display:none containers, canvas fallback content)

twilco opened the pull request:
#### 7662529ead2b8c70358f869c2e295d23cbc9861a
<pre>
AX: WebKit sometimes exposes duplicate text for unrendered text nodes (e.g. labels inside display:none containers, canvas fallback content)
<a href="https://bugs.webkit.org/show_bug.cgi?id=290486">https://bugs.webkit.org/show_bug.cgi?id=290486</a>
<a href="https://rdar.apple.com/147969486">rdar://147969486</a>

Reviewed by Ryosuke Niwa.

In AccessibilityNodeObject::textUnderElement, we used Text::wholeText rather than Text::data. The former can gather
text from sibiling text nodes, meaning we could doubly-expose text. Text::data only returns text belonging to the
given node, not any siblings, which is the behavior we want, and now have with this commit.

Beyond affecting canvas fallback content and diplay:none labels (which still gather descendant text despite not being
rendered), this bug also resulted in otherwise empty, unrendered text elements to be exposed, since they scooped up
their siblings text.

The test added in this commit also exposed an ITM-specific bug where we didn&apos;t update the accessibility tree and any
label text for node-only subtrees. This is fixed by changing ContainerNode::childrenChanged to call AXObjectCache::childrenChanged(Node&amp;),
which now only does anything when the given node has no renderer. If it does have a renderer, we&apos;ll already get the
appropriate children-changed notifications from the render tree.

A few drive-by fixes were made to pass references to a few AXObjectCache methods, as all callsites have already performed
null-checks. This improves performance a tiny bit.

* LayoutTests/accessibility/duplicate-hidden-label-text-expected.txt: Added.
* LayoutTests/accessibility/duplicate-hidden-label-text.html: Added.
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* LayoutTests/platform/ios/accessibility/duplicate-hidden-label-text-expected.txt: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::handleAttributeChange):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::textUnderElement const):
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::AccessibilityObject::setLastPresentedTextPrediction):
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::childrenChanged):
* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::optionElementChildrenChanged):
(WebCore::HTMLSelectElement::setRecalcListItems):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::removeFromAXObjectCache):
(WebCore::LocalFrameView::willRemoveScrollbar):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::setHasScrollbarInternal):
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::willRemoveScrollbar):
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::destroyScrollbar):
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::destroyScrollbar):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::willBeDestroyed):
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::willBeDestroyed):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::destroyScrollbar):

Canonical link: <a href="https://commits.webkit.org/292818@main">https://commits.webkit.org/292818@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/abd4d23c4daee16d8d54c8eb389f349200dd9214

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97072 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16695 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6903 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102153 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47597 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99117 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16987 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25145 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73950 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31163 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100075 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12819 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54287 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12571 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5649 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46927 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82638 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5727 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104175 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24146 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82998 "Passed tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/24517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83900 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82396 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20752 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17651 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24111 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29262 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23934 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27246 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25507 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
```diff
@@ -150,7 +150,7 @@
     if (state == CompositionState::Ended && !lastPresentedTextPrediction().text.isEmpty()) {
         auto* nodeText = dynamicDowncast<Text>(previousCompositionNode);
-        String previousCompositionNodeText = nodeText ? nodeText->wholeText() : String();
+        String previousCompositionNodeText = nodeText ? nodeText->data() : String();
```
So this is the fix for the bug. I wish pointer -> reference refactoring, etc... were separate.

twilco replied:
> I wish pointer -> reference refactoring, etc... were separate.

Yeah, you're right. Walked back most of those changes. Did have to keep a few other things not related to the `wholeText()` to `data()` switch in order for this new test to pass, as described in the commit message, but this is more clean now.

twilco committed:
AX: WebKit sometimes exposes duplicate text for unrendered text nodes (e.g. labels inside display:none containers, canvas fallback content)
https://bugs.webkit.org/show_bug.cgi?id=290486
rdar://147969486

Reviewed by Ryosuke Niwa.

In AccessibilityNodeObject::textUnderElement, we used Text::wholeText rather than Text::data. The former can gather
text from sibiling text nodes, meaning we could doubly-expose text. Text::data only returns text belonging to the
given node, not any siblings, which is the behavior we want, and now have with this commit.

Beyond affecting canvas fallback content and diplay:none labels (which still gather descendant text despite not being
rendered), this bug also resulted in otherwise empty, unrendered text elements to be exposed, since they scooped up
their siblings text.

The test added in this commit also exposed an ITM-specific bug where we didn't update the accessibility tree and any
label text for node-only subtrees. This is fixed by changing ContainerNode::childrenChanged to call AXObjectCache::childrenChanged(Node&),
which now only does anything when the given node has no renderer. If it does have a renderer, we'll already get the
appropriate children-changed notifications from the render tree.

A few drive-by fixes were made to pass references to a few AXObjectCache methods, as all callsites have already performed
null-checks. This improves performance a tiny bit.

* LayoutTests/accessibility/duplicate-hidden-label-text-expected.txt: Added.
* LayoutTests/accessibility/duplicate-hidden-label-text.html: Added.
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* LayoutTests/platform/ios/accessibility/duplicate-hidden-label-text-expected.txt: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::handleAttributeChange):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::textUnderElement const):
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::AccessibilityObject::setLastPresentedTextPrediction):
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::childrenChanged):
* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::optionElementChildrenChanged):
(WebCore::HTMLSelectElement::setRecalcListItems):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::removeFromAXObjectCache):
(WebCore::LocalFrameView::willRemoveScrollbar):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::setHasScrollbarInternal):
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::willRemoveScrollbar):
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::destroyScrollbar):
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::destroyScrollbar):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::willBeDestroyed):
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::willBeDestroyed):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::destroyScrollbar):

Canonical link: https://commits.webkit.org/292818@main
